File: 프로그래머스/lv3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    answer = 0
    graph = [[] for _ in range(n+1)] # 연결된 노드 정보 그래프
    distance = [-1] * (n+1) # 각 노드의 최단거리 리스트
    
    # 연결된 노드 정보 추가
    for e in edge:
        graph[e[0]].append(e[1])
        graph[e[1]].append(e[0])
    if(len(graph[1]) == 0):
        return 0
    queue = deque([1]) # BFS를 위한 큐, 출발 노드는 1
    distance[1] = 0 # 출발 노드의 최단거리 0으로 
    
    
    # BFS 수행
    while queue:
        now = queue.popleft() # 현재 노드
         
        # 현재 노드에서 이동할 수 있는 모든 노드 확인
        for i in graph[now]:
            if distance[i] == -1: # 아직 방문하지 않은 노드라면
                queue.append(i) # 큐에 추가
                distance[i] = distance[now] + 1 # 최단거리 갱신
    
    # 가장 멀리 떨어진 노드 개수 구하기
    for d in distance:
        if d == max(distance):
            answer += 1
            
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    static List<Integer>[] graph;
    
    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        Deque<int[]> deque = new ArrayDeque<>();
        int[] distance = new int[n+1];
        Arrays.fill(distance, -1);
        distance[1] = 0;
        
        graph = new ArrayList[n+1];
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int i = 0; i < edge.length; i++) {
            graph[edge[i][0]].add(edge[i][1]);
            graph[edge[i][1]].add(edge[i][0]);
        }
        
        for (Integer node : graph[1]) {
            deque.add(new int[] {1, node});
        }
        
        int maxDistance = 0;
        while(!deque.isEmpty()) {
            int[] dq = deque.poll();
    
            if (distance[dq[1]] != -1) continue;
            
            distance[dq[1]] = distance[dq[0]] + 1;
            maxDistance = Math.max(maxDistance, distance[dq[1]]);
            
            for (Integer node : graph[dq[1]]) {
                if (distance[node] == -1) 
                    deque.add(new int[] {dq[1], node});
            }
        }
     
        if (maxDistance > 0) {
            for (int i = 1; i <=n; i++) {
                if (distance[i] == maxDistance) 
                    answer += 1;
            }
        }
        
        return answer;
    }
}